### PR TITLE
Bnb/gcn

### PR DESCRIPTION
**Pull request by bnb32**
Working GCN with batching. Added documentation workflow.

### CODE REVIEW

File: docs/source/conf.py [hunk 1]
```diff
@@ -0,0 +1,206 @@
...
+import os
+import sphinx_rtd_theme
+import sys
+sys.path.insert(0, os.path.abspath('../../'))
+
+# -- Project information -----------------------------------------------------
+
+project = 'rl_equation_solver'
+copyright = '2023'
+author = "Brandon Benton, Kevin O'Keefe"
```
Comment (Khev) [inline docs/source/conf.py:27]: O'Keefe --> O'Keeffe

File: rl_equation_solver/utilities/loss.py [hunk 2]
```diff
@@ -0,0 +1,18 @@
+"""Collection of loss functions"""
+import torch
+
+
+class LossMixin:
+    """Mixin class with collection of loss functions"""
```
Comment (Khev) [inline rl_equation_solver/utilities/loss.py:6]: Mixin -- what does this mean?

File: rl_equation_solver/utilities/loss.py [hunk 3]
```diff
@@ -0,0 +1,18 @@
...
+
+    def huber_loss(self, x, y, delta=1.0):
+        """Huber loss"""
```
Comment (Khev) [inline rl_equation_solver/utilities/loss.py:9]: Maybe define the huber loss:

Huber loss, also known as Smooth Mean Absolute Error, is a loss function used in various machine learning and optimization problems, particularly in regression tasks. It combines the properties of both Mean Squared Error (MSE) and Mean Absolute Error (MAE) loss functions, providing a balance between the two.

L(y, f(x)) = 0.5 * (y - f(x))^2                 if |y - f(x)| <= delta
            delta * |y - f(x)| - 0.5 * delta^2  otherwise

File: rl_equation_solver/utilities/reward.py [hunk 4]
```diff
@@ -0,0 +1,91 @@
...
+from abc import abstractmethod
+import numpy as np
+
+
+class RewardMixin:
+    """Reward function collection"""
+
+    @abstractmethod
+    def expression_complexity(self, state):
+        """Get the graph / expression complexity for a given state"""
```
Comment (Khev) [inline rl_equation_solver/utilities/reward.py:11]: where graph complexity := number of nodes + number of edges. Example:

The graph for ax + b = 0 is:

        =
       /   \
      add  0
    /   \
   mult  b
 /      \
a     x

which has 7 nodes + 6 edges = complexity of 13

Comment (Khev) [reply rl_equation_solver/utilities/reward.py:11]: Shoot, the graph above isn't rendering

File: rl_equation_solver/utilities/reward.py [hunk 5]
```diff
@@ -0,0 +1,91 @@
+"""Collection of reward functions"""
```
Comment (Khev) [inline rl_equation_solver/utilities/reward.py:1]: Nice. Good idea to collect all the rewards in a separate file

File: rl_equation_solver/utilities/utilities.py [hunk 6]
```diff
@@ -0,0 +1,316 @@
...
+import numpy as np
+import scipy.sparse as sp
+from collections import namedtuple
+import torch
+from torch_geometric.utils.convert import from_networkx
+import networkx as nx
+from networkx.readwrite import json_graph
+from networkx.drawing.nx_pydot import graphviz_layout
+
+
```
Comment (Khev) [inline rl_equation_solver/utilities/utilities.py:11]: Add comment: For agents with experience replay (not all agents have this)

File: rl_equation_solver/utilities/utilities.py [hunk 7]
```diff
@@ -0,0 +1,316 @@
...
+    """
+    graph = get_json_graph(expr)
+    node_features = get_node_features(graph, feature_dict)
+    for i, node in enumerate(list(graph.nodes)):
+        graph.nodes[node]['x'] = node_features[i]
+    return graph
+
+
+def get_node_features(graph, feature_dict):
+    """Get node features from feature dictionary"""
```
Comment (Khev) [inline rl_equation_solver/utilities/utilities.py:235]: Example of node features: we map the operations and terms to integeters: {add: 0, subt: 1, .. }

**Review by Khev — APPROVED** [review]
lgtm. some comments here and there

File: rl_equation_solver/utilities/loss.py (hunk 2 above)
Comment (bnb32) [reply rl_equation_solver/utilities/loss.py:6]: General term for classes that aren't meant to be super classes just supposed to "mix in" a collection of methods.
